### DISCUSSION TITLE
How to build `liboqs` for Arm Cortex-M4?

### DISCUSSION THREAD

**Discussion by tarun14110** [open-quantum-safe/liboqs#1457] [Q&A]
How to build `liboqs` for Arm Cortex-M4 processor on MacOS? I am trying to build using
`cmake -GNinja ..  -DCMAKE_TOOLCHAIN_FILE=../.CMake/toolchain_armel.cmake`. I have added `set(CMAKE_C_COMPILER arm-none-eabi-gcc)` and `set(CMAKE_CROSSCOMPILING_EMULATOR "qemu-system-arm;-L;/opt/homebrew/bin/")` in toolchain_armel.cmake

Here is the output of the command.

> -- The C compiler identification is GNU 10.3.1
> -- The ASM compiler identification is GNU
> -- Found assembler: /Applications/ARM/bin/arm-none-eabi-gcc
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - failed
> -- Check for working C compiler: /Applications/ARM/bin/arm-none-eabi-gcc
> -- Check for working C compiler: /Applications/ARM/bin/arm-none-eabi-gcc - broken
> CMake Error at /opt/homebrew/Cellar/cmake/3.26.3/share/cmake/Modules/CMakeTestCCompiler.cmake:67 (message):
>   The C compiler
> 
>     "/Applications/ARM/bin/arm-none-eabi-gcc"
> 
>   is not able to compile a simple test program.
> 
>   It fails with the following output:
> 
>     Change Dir: /Users/.../liboqs/build/CMakeFiles/CMakeScratch/TryCompile-airKbw
>     
>     Run Build Command(s):/opt/homebrew/bin/ninja -v cmTC_bcc40 && [1/2] /Applications/ARM/bin/arm-none-eabi-gcc    -o CMakeFiles/cmTC_bcc40.dir/testCCompiler.c.o -c /Users/.../liboqs/build/CMakeFiles/CMakeScratch/TryCompile-airKbw/testCCompiler.c
>     [2/2] : && /Applications/ARM/bin/arm-none-eabi-gcc   CMakeFiles/cmTC_bcc40.dir/testCCompiler.c.o -o cmTC_bcc40   && :
>     FAILED: cmTC_bcc40 
>     : && /Applications/ARM/bin/arm-none-eabi-gcc   CMakeFiles/cmTC_bcc40.dir/testCCompiler.c.o -o cmTC_bcc40   && :
>     /Applications/ARM/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ARM/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/lib/libc.a(lib_a-exit.o): in function `exit':
>     exit.c:(.text.exit+0x2c): undefined reference to `_exit'
>     collect2: error: ld returned 1 exit status
>     ninja: build stopped: subcommand failed.
>     
>     
> 
>   
> 
>   CMake will not be able to correctly generate this project.
> Call Stack (most recent call first):
>   CMakeLists.txt:22 (project)

**Comment by baentsch** [comment]
I personally have no experience with cross-compiling (let alone on MacOS) but this error message seems like a standard/non-OQS-specific installation/setup problem (just looking at stackoverflow.com...). 

Also, have you tried with the `armhf` toolchain file instead? Maybe that crosscompiler works better ...

Final suggestion: Don't use an OSX crosscompile environment, but one that we know works, namely a Linux image (via docker). See for example https://github.com/open-quantum-safe/liboqs/blob/9f912c957bfe7f4b894aa9661168a310e8dd1a58/.github/workflows/linux.yml#L133-L142
